add definition of IS25LP016 device

diff -Naru ../original/flashrom-v1.2.1/flashchips.c flashrom-v1.2.1/flashchips.c
--- ../original/flashrom-v1.2.1/flashchips.c	2025-04-25 02:51:33.584437556 +0000
+++ flashrom-v1.2.1/flashchips.c	2025-04-25 04:06:56.803581437 +0000
@@ -6906,6 +6906,47 @@
 
 	{
 		.vendor		= "ISSI",
+		.name		= "IS25LP016",
+		.bustype	= BUS_SPI,
+		.manufacture_id = ISSI_ID_SPI,
+		.model_id	= ISSI_IS25LP016,
+		.total_size	= 2048,
+		.page_size	= 256,
+		/* OTP: 1024B total; read 0x48; write 0x42 */
+		.feature_bits	= FEATURE_WRSR_WREN | FEATURE_OTP,
+		.tested		= TEST_UNTESTED,
+		.probe		= probe_spi_rdid,
+		.probe_timing	= TIMING_ZERO,
+		.block_erasers	=
+		{
+			{
+				.eraseblocks = { {4 * 1024, 512} },
+				.block_erase = spi_block_erase_20,
+			}, {
+				.eraseblocks = { {4 * 1024, 512} },
+				.block_erase = spi_block_erase_d7,
+			}, {
+				.eraseblocks = { {32 * 1024, 64} },
+				.block_erase = spi_block_erase_52,
+			}, {
+				.eraseblocks = { {64 * 1024, 32} },
+				.block_erase = spi_block_erase_d8,
+			}, {
+				.eraseblocks = { {2 * 1024 * 1024, 1} },
+				.block_erase = spi_block_erase_60,
+			}, {
+				.eraseblocks = { {2 * 1024 * 1024, 1} },
+				.block_erase = spi_block_erase_c7,
+			}
+		},
+		.unlock		= spi_disable_blockprotect_bp3_srwd,
+		.write		= spi_chip_write_256,
+		.read		= spi_chip_read,
+		.voltage	= {2300, 3600},
+	},
+
+	{
+		.vendor		= "ISSI",
 		.name		= "IS25LP064",
 		.bustype	= BUS_SPI,
 		.manufacture_id	= ISSI_ID_SPI,
diff -Naru ../original/flashrom-v1.2.1/flashchips.h flashrom-v1.2.1/flashchips.h
--- ../original/flashrom-v1.2.1/flashchips.h	2025-04-25 02:51:33.797696896 +0000
+++ flashrom-v1.2.1/flashchips.h	2025-04-25 04:06:57.123219487 +0000
@@ -456,6 +456,7 @@
 
 #define ISSI_ID			0xD5	/* ISSI Integrated Silicon Solutions, see also PMC. */
 #define ISSI_ID_SPI		0x9D	/* ISSI ID used for SPI flash, see also PMC_ID_NOPREFIX */
+#define ISSI_IS25LP016		0x6015
 #define ISSI_IS25LP064		0x6017
 #define ISSI_IS25LP128		0x6018
 #define ISSI_IS25LP256		0x6019
